File: 06-1-Retrofit1App/app/src/main/java/com/example/retrofit1app/MainViewModel.java
package com.example.retrofit1app;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.retrofit1app.service.AnimeQuote;
import com.example.retrofit1app.service.AnimeQuoteRepository;


public class MainViewModel extends ViewModel {
    final MutableLiveData<Boolean> showLoading = new MutableLiveData<>(false);
    final MutableLiveData<String> quote = new MutableLiveData<>();

    public MutableLiveData<Boolean> getShowLoading() {
        return showLoading;
    }

    public MutableLiveData<String> getQuote() {
        return quote;
    }

    public void newQuote() {
        quote.setValue("");
        showLoading.setValue(true);
        AnimeQuoteRepository.getQuote(new AnimeQuoteRepository.AnimeQuoteCallback() {
            @Override
            public void onSuccess(AnimeQuote animeQuote) {
                showLoading.setValue(false);
                quote.setValue(animeQuote.getQuote());
            }

            @Override
            public void onError(String errorMessage) { }
        });
    }
}

File: 06-1-Retrofit1App_kt2/app/src/main/java/com/example/retrofit1app_kt2/MainViewModel.kt
package com.example.retrofit1app_kt2

import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.lifecycle.viewModelScope
import com.example.retrofit1app_kt2.service.RetrofitInstance
import kotlinx.coroutines.launch

class MainViewModel : ViewModel() {
    private val api = RetrofitInstance.api

    var quote by mutableStateOf("")
    var loading by mutableStateOf(false)

    fun newQuote() {
        quote = ""
        loading = true

        viewModelScope.launch {
            try {
                val randomQuote = api.getRandomQuote()
                quote = randomQuote.quote
            } catch (e: Exception) {
                quote = "Error"
            } finally {
                loading = false
            }
        }
    }
}